File: go/parser.go
package slon

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"
	"unicode"
)

// Parse converts a SLON string into Go values.
// Objects become map[string]any, arrays []any, datetimes time.Time in UTC.
func Parse(input string) (any, error) {
	p := &parser{text: input}
	p.skipWhitespace()
	value, err := p.parseValue()
	if err != nil {
		return nil, err
	}
	p.skipWhitespace()
	if !p.end() {
		return nil, fmt.Errorf("unexpected trailing content at position %d", p.pos)
	}
	return value, nil
}

type parser struct {
	text string
	pos  int
}

func (p *parser) end() bool {
	return p.pos >= len(p.text)
}

func (p *parser) peek() (rune, bool) {
	if p.end() {
		return 0, false
	}
	return rune(p.text[p.pos]), true
}

func (p *parser) advance(n int) {
	p.pos += n
}

func (p *parser) skipWhitespace() {
	for !p.end() {
		if !unicode.IsSpace(rune(p.text[p.pos])) {
			break
		}
		p.pos++
	}
}

func (p *parser) parseValue() (any, error) {
	p.skipWhitespace()
	ch, ok := p.peek()
	if !ok {
		return nil, fmt.Errorf("unexpected end of input")
	}
	switch ch {
	case '(':
		return p.parseObject()
	case '[':
		return p.parseArray()
	case '\'', '"':
		return p.parseQuotedString()
	case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		if dt, consumed, err := p.tryParseDateTime(); err != nil {
			return nil, err
		} else if consumed {
			return dt, nil
		}
		return p.parseNumber()
	default:
		if keyword, consumed := p.matchKeyword("true"); consumed {
			return keyword, nil
		}
		if keyword, consumed := p.matchKeyword("false"); consumed {
			return keyword, nil
		}
		if keyword, consumed := p.matchKeyword("null"); consumed {
			return keyword, nil
		}
		return p.parseUnquotedString()
	}
}

func (p *parser) parseObject() (map[string]any, error) {
	p.advance(1) // skip '('
	result := make(map[string]any)
	p.skipWhitespace()
	if ch, ok := p.peek(); ok && ch == ')' {
		p.advance(1)
		return result, nil
	}
	for {
		p.skipWhitespace()
		key, err := p.parseStringLike()
		if err != nil {
			return nil, err
		}
		p.skipWhitespace()
		if ch, ok := p.peek(); !ok || ch != ':' {
			return nil, fmt.Errorf("expected ':' after key at position %d", p.pos)
		}
		p.advance(1)
		value, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		result[key] = value
		p.skipWhitespace()
		ch, ok := p.peek()
		if !ok {
			return nil, fmt.Errorf("unterminated object")
		}
		if ch == ',' {
			p.advance(1)
			continue
		}
		if ch == ')' {
			p.advance(1)
			break
		}
		return nil, fmt.Errorf("expected ',' or ')' at position %d", p.pos)
	}
	return result, nil
}

func (p *parser) parseArray() ([]any, error) {
	p.advance(1) // skip '['
	var result []any
	p.skipWhitespace()
	if ch, ok := p.peek(); ok && ch == ']' {
		p.advance(1)
		return result, nil
	}
	for {
		value, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		result = append(result, value)
		p.skipWhitespace()
		ch, ok := p.peek()
		if !ok {
			return nil, fmt.Errorf("unterminated array")
		}
		if ch == '|' {
			p.advance(1)
			continue
		}
		if ch == ']' {
			p.advance(1)
			break
		}
		return nil, fmt.Errorf("expected '|' or ']' at position %d", p.pos)
	}
	return result, nil
}

func (p *parser) parseStringLike() (string, error) {
	if ch, ok := p.peek(); ok && (ch == '\'' || ch == '"') {
		return p.parseQuotedString()
	}
	str, err := p.parseUnquotedString()
	return str, err
}

func (p *parser) parseQuotedString() (string, error) {
	ch, _ := p.peek()
	quote := byte(ch)
	p.advance(1)
	var builder strings.Builder
	for !p.end() {
		current := p.text[p.pos]
		if current == quote {
			p.advance(1)
			return builder.String(), nil
		}
		if current == '\\' {
			if p.pos+1 >= len(p.text) {
				return "", fmt.Errorf("invalid escape at position %d", p.pos)
			}
			next := p.text[p.pos+1]
			switch next {
			case '"', '\'', '\\', '/':
				builder.WriteByte(next)
			case 'b':
				builder.WriteByte('\b')
			case 'f':
				builder.WriteByte('\f')
			case 'n':
				builder.WriteByte('\n')
			case 'r':
				builder.WriteByte('\r')
			case 't':
				builder.WriteByte('\t')
			case 'u':
				if p.pos+5 >= len(p.text) {
					return "", fmt.Errorf("invalid unicode escape at position %d", p.pos)
				}
				hex := p.text[p.pos+2 : p.pos+6]
				r, err := strconv.ParseInt(hex, 16, 32)
				if err != nil {
					return "", fmt.Errorf("invalid unicode escape at position %d", p.pos)
				}
				builder.WriteRune(rune(r))
				p.advance(4)
			default:
				return "", fmt.Errorf("unknown escape at position %d", p.pos)
			}
			p.advance(2)
			continue
		}
		builder.WriteByte(current)
		p.advance(1)
	}
	return "", fmt.Errorf("unterminated string literal")
}

func (p *parser) parseUnquotedString() (string, error) {
	start := p.pos
	for !p.end() {
		ch := p.text[p.pos]
		if isDelimiter(ch) || unicode.IsSpace(rune(ch)) {
			break
		}
		p.pos++
	}
	raw := strings.TrimSpace(p.text[start:p.pos])
	if raw == "" {
		return "", fmt.Errorf("empty string at position %d", start)
	}
	return raw, nil
}

func (p *parser) parseNumber() (any, error) {
	start := p.pos
	for !p.end() {
		ch := p.text[p.pos]
		if !(ch == '+' || ch == '-' || ch == 'e' || ch == 'E' || ch == '.' || (ch >= '0' && ch <= '9')) {
			break
		}
		p.pos++
	}
	number := p.text[start:p.pos]
	if number == "" {
		return nil, fmt.Errorf("invalid number at position %d", start)
	}
	if p.pos < len(p.text) {
		if ch := p.text[p.pos]; !isDelimiter(ch) && !unicode.IsSpace(rune(ch)) {
			return nil, fmt.Errorf("invalid number boundary at position %d", p.pos)
		}
	}
	if strings.ContainsAny(number, ".eE") {
		value, err := strconv.ParseFloat(number, 64)
		if err != nil || math.IsNaN(value) || math.IsInf(value, 0) {
			return nil, fmt.Errorf("invalid float at position %d", start)
		}
		return value, nil
	}
	value, err := strconv.ParseInt(number, 10, 64)
	if err == nil {
		return value, nil
	}
	unsigned, errUnsigned := strconv.ParseUint(number, 10, 64)
	if errUnsigned != nil {
		return nil, fmt.Errorf("invalid integer at position %d", start)
	}
	return unsigned, nil
}

func (p *parser) tryParseDateTime() (time.Time, bool, error) {
	if p.pos+23 > len(p.text) {
		return time.Time{}, false, nil
	}
	candidate := p.text[p.pos : p.pos+23]
	layout := "2006-01-02/15:04:05.000"
	if p.pos+23 < len(p.text) {
		next := p.text[p.pos+23]
		if !isDelimiter(next) && !unicode.IsSpace(rune(next)) {
			return time.Time{}, false, nil
		}
	}
	t, err := time.ParseInLocation(layout, candidate, time.UTC)
	if err != nil {
		return time.Time{}, false, nil
	}
	p.pos += 23
	return t, true, nil
}

func (p *parser) matchKeyword(keyword string) (any, bool) {
	if strings.HasPrefix(p.text[p.pos:], keyword) {
		end := p.pos + len(keyword)
		if end == len(p.text) || isDelimiter(p.text[end]) || unicode.IsSpace(rune(p.text[end])) {
			p.pos = end
			switch keyword {
			case "true":
				return true, true
			case "false":
				return false, true
			case "null":
				return nil, true
			}
		}
	}
	return nil, false
}

func isDelimiter(ch byte) bool {
	switch ch {
	case ':', ',', '(', ')', '[', ']', '|':
		return true
	default:
		return false
	}
}

File: python/slon_python/parser.py
import io
import re
from datetime import datetime, timezone
from typing import Any, Dict, Iterable, List, Union


class SLONDecodeError(ValueError):
    """Raised when parsing fails."""


_DATETIME_RE = re.compile(
    r"""
    (?P<year>\d{4})-
    (?P<month>\d{2})-
    (?P<day>\d{2})/
    (?P<hour>\d{2}):
    (?P<minute>\d{2}):
    (?P<second>\d{2})\.
    (?P<millisecond>\d{3})
    """,
    re.VERBOSE,
)

_NUMBER_RE = re.compile(r"-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?")
_DELIMS = {":", ",", "(", ")", "[", "]", "|"}
_WS = {" ", "\t", "\r", "\n"}


class _Parser:
    def __init__(self, text: str):
        self.text = text
        self.length = len(text)
        self.pos = 0

    def parse(self) -> Any:
        self._skip_ws()
        value = self._parse_value()
        self._skip_ws()
        if self.pos != self.length:
            raise SLONDecodeError(f"Unexpected trailing content at position {self.pos}")
        return value

    def _peek(self) -> Union[str, None]:
        return self.text[self.pos] if self.pos < self.length else None

    def _advance(self, count: int = 1) -> None:
        self.pos += count

    def _skip_ws(self) -> None:
        while self.pos < self.length and self.text[self.pos] in _WS:
            self.pos += 1

    def _expect(self, char: str) -> None:
        if self._peek() != char:
            raise SLONDecodeError(f"Expected '{char}' at position {self.pos}")
        self._advance()

    def _parse_value(self) -> Any:
        self._skip_ws()
        ch = self._peek()
        if ch is None:
            raise SLONDecodeError("Unexpected end of input")
        if ch == "(":
            return self._parse_object()
        if ch == "[":
            return self._parse_array()
        if ch in {"'", '"'}:
            return self._parse_quoted_string()
        if ch in "-0123456789":
            dt = self._try_parse_datetime()
            if dt is not None:
                return dt
            return self._parse_number()
        if self._match_keyword("true"):
            return True
        if self._match_keyword("false"):
            return False
        if self._match_keyword("null"):
            return None
        return self._parse_unquoted_string()

    def _match_keyword(self, keyword: str) -> bool:
        end = self.pos + len(keyword)
        if self.text[self.pos:end] == keyword:
            boundary = end == self.length or self.text[end] in _DELIMS or self.text[end] in _WS
            if boundary:
                self.pos = end
                return True
        return False

    def _parse_object(self) -> Dict[str, Any]:
        self._expect("(")
        obj: Dict[str, Any] = {}
        self._skip_ws()
        if self._peek() == ")":
            self._advance()
            return obj
        while True:
            self._skip_ws()
            key = self._parse_string_like()
            self._skip_ws()
            self._expect(":")
            value = self._parse_value()
            obj[key] = value
            self._skip_ws()
            ch = self._peek()
            if ch == ",":
                self._advance()
                continue
            if ch == ")":
                self._advance()
                break
            raise SLONDecodeError(f"Expected ',' or ')' at position {self.pos}")
        return obj

    def _parse_array(self) -> List[Any]:
        self._expect("[")
        items: List[Any] = []
        self._skip_ws()
        if self._peek() == "]":
            self._advance()
            return items
        while True:
            value = self._parse_value()
            items.append(value)
            self._skip_ws()
            ch = self._peek()
            if ch == "|":
                self._advance()
                continue
            if ch == "]":
                self._advance()
                break
            raise SLONDecodeError(f"Expected '|' or ']' at position {self.pos}")
        return items

    def _parse_quoted_string(self) -> str:
        quote = self._peek()
        assert quote in {"'", '"'}
        self._advance()
        chars: List[str] = []
        while True:
            ch = self._peek()
            if ch is None:
                raise SLONDecodeError("Unterminated string literal")
            if ch == quote:
                self._advance()
                break
            if ch == "\\":
                self._advance()
                esc = self._peek()
                if esc is None:
                    raise SLONDecodeError("Invalid escape at end of string")
                chars.append(self._parse_escape(esc))
                self._advance()
            else:
                chars.append(ch)
                self._advance()
        return "".join(chars)

    def _parse_escape(self, esc: str) -> str:
        mapping = {
            '"': '"',
            "'": "'",
            "\\": "\\",
            "/": "/",
            "b": "\b",
            "f": "\f",
            "n": "\n",
            "r": "\r",
            "t": "\t",
        }
        if esc in mapping:
            return mapping[esc]
        if esc == "u":
            start = self.pos + 1
            end = start + 4
            if end > self.length:
                raise SLONDecodeError("Invalid unicode escape sequence")
            hex_part = self.text[start:end]
            if not re.fullmatch(r"[0-9a-fA-F]{4}", hex_part):
                raise SLONDecodeError("Invalid unicode escape sequence")
            self.pos = end - 1
            return chr(int(hex_part, 16))
        raise SLONDecodeError(f"Unknown escape '\\{esc}'")

    def _parse_unquoted_string(self) -> str:
        start = self.pos
        while self.pos < self.length:
            ch = self.text[self.pos]
            if ch in _DELIMS or ch in _WS:
                break
            self.pos += 1
        value = self.text[start:self.pos].strip()
        if not value:
            raise SLONDecodeError(f"Empty string value at position {start}")
        return value

    def _parse_string_like(self) -> str:
        ch = self._peek()
        if ch in {"'", '"'}:
            return self._parse_quoted_string()
        return self._parse_unquoted_string()

    def _parse_number(self) -> Union[int, float]:
        match = _NUMBER_RE.match(self.text, self.pos)
        if not match:
            raise SLONDecodeError(f"Invalid number at position {self.pos}")
        end = match.end()
        boundary = end == self.length or self.text[end] in _DELIMS or self.text[end] in _WS
        if not boundary:
            raise SLONDecodeError(f"Invalid number boundary at position {end}")
        number = match.group()
        self.pos = end
        if "." in number or "e" in number or "E" in number:
            return float(number)
        return int(number)

    def _try_parse_datetime(self) -> Union[datetime, None]:
        match = _DATETIME_RE.match(self.text, self.pos)
        if not match:
            return None
        end = match.end()
        boundary = end == self.length or self.text[end] in _DELIMS or self.text[end] in _WS
        if not boundary:
            return None
        groups = {k: int(v) for k, v in match.groupdict().items()}
        try:
            dt = datetime(
                groups["year"],
                groups["month"],
                groups["day"],
                groups["hour"],
                groups["minute"],
                groups["second"],
                groups["millisecond"] * 1000,
                tzinfo=timezone.utc,
            )
        except ValueError as exc:
            raise SLONDecodeError(f"Invalid datetime literal at position {self.pos}") from exc
        self.pos = end
        return dt


def loads(text: str) -> Any:
    """Parse a SLON string into Python values."""
    return _Parser(text).parse()


def load(stream: io.TextIOBase) -> Any:
    """Parse SLON content from a file-like object."""
    return loads(stream.read())


def dumps(value: Any) -> str:
    """Serialize Python values into SLON."""
    return _stringify(value)


def dump(value: Any, stream: io.TextIOBase) -> None:
    """Serialize Python values as SLON into a file-like object."""
    stream.write(dumps(value))


def _stringify(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return repr(value)
    if isinstance(value, datetime):
        dt = value.astimezone(timezone.utc)
        return dt.strftime("%Y-%m-%d/%H:%M:%S.%f")[:-3]
    if isinstance(value, str):
        return _format_string(value)
    if isinstance(value, dict):
        items: List[str] = []
        for key in sorted(value.keys()):
            items.append(f"{_format_key(key)}: {_stringify(value[key])}")
        return f"({', '.join(items)})"
    if isinstance(value, (list, tuple)):
        return f"[{' | '.join(_stringify(v) for v in value)}]"
    raise TypeError(f"Unsupported type: {type(value)!r}")


def _format_key(key: str) -> str:
    if not key:
        return f"'{key}'"
    if any(ch in _DELIMS or ch in {"'", '"'} or ch.isspace() for ch in key):
        return _format_string(key)
    return key


def _format_string(value: str) -> str:
    escaped = value.replace("\\", "\\\\").replace("'", "\\'")
    escaped = escaped.replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")
    return f"'{escaped}'"
